### PR TITLE
Adds facebook omniauth login.

### PR DESCRIPTION
**Pull request by coreycondardo**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -6,16 +6,21 @@ def new
...
+
   def create
-    user = User.find_by(email: params[:email])
-    if user && user.authenticate(params[:password])
-      session[:user_id] = user.id 
+    if omni_auth_hash
+      user, error_message  = auth_omni_auth_user(omni_auth_hash)
+    else
+      user, error_message = auth_password_user(params)
+    end
```
Comment (ptrckbrwn) [inline app/controllers/sessions_controller.rb]: Could be rewritten as:

```ruby
user, error_message = if omni_auth_hash
                        auth_omni_auth_user(omni_auth_hash)
                      else
                        auth_password_user(params)
                      end
```

I think this is a little clearer that you're performing this `if` to set these two variables.

Comment (coreycondardo) [inline app/controllers/sessions_controller.rb]: Ah totally, good call.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -25,4 +30,30 @@ def destroy
...
+  private
+
+  def omni_auth_hash
+    request.env['omniauth.auth']
+  end
+
+  def auth_password_user(params)
+    user = User.find_by(email: params[:email])
+
+    unless !!(user && user.authenticate(params[:password]))
```
Comment (ptrckbrwn) [inline app/controllers/sessions_controller.rb]: `!!` is a no-op here, because `&&` already returns a boolean.

**Comment by coreycondardo** [comment]
Should add a new column to track if a user is oauthed.

File: app/models/user.rb [hunk 3]
```diff
@@ -11,4 +11,18 @@ def generate_token
...
   end
+
+  def self.find_or_create_from_omni_auth_hash(omni_auth_hash)
+    email = omni_auth_hash.fetch(:info, {}).fetch(:email, nil)
+    user = User.find_by_email(email)
+
+    if user.nil?
+      user = User.create(
+        :email => email,
+        :password => SecureRandom.hex(10)
```
Comment (ptrckbrwn) [inline app/models/user.rb]: I'm assuming this is because Oauth users don't need a password, but you need to satiate a database constraint, correct?

**Comment by coreycondardo** [comment]
Add specs!

File: app/models/user.rb (hunk 3 above)
Comment (coreycondardo) [inline app/models/user.rb]: Sadly, yes. They can reset pw if they wanna use email instead, assuming the email the same as their fb email.

**Comment by ptrckbrwn** [comment]
# Micro

Looks clean, easy to read, not overly complex. :100: 

# Macro

The concerns of a password authenticated user is different than a oauth authenticated user. As your app grows you might consider factoring these out into two different classes that `User` has a polymorphic relationship to. That way you could login a user like:

```ruby
user, error_message = if omni_auth_hash
                        OmniAuthUser.login(omni_auth_hash)
                      else
                        PasswordAuthUser.login(password)
                      end
```

**Comment by coreycondardo** [comment]
Totally, I think if there is anymore work or conditionals introduced it makes sense. The decisions trees there already scream "new class" but I was trying to avoid it for the sake of shipping it. Appreciate the review.

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (coreycondardo) [inline app/controllers/sessions_controller.rb]: It seems like if `user` is `nil`, it returns nil since it short circuits there.

**Pull request merged**